src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             }
 
             // We've probed to find an empty spot, add to hash
-            if (!TryWriteValueToLocation(value, hashTableLocal, tableIndex))
+            if (!TryWriteSentinelToLocation(hashTableLocal, tableIndex))
             {
+                // After finding the empty spot, it was taken by some other thread
                 Interlocked.Decrement(ref _reserve);
                 return null;
             }
 
             // Now that we've written to the local array, find out if that array has been
             // replaced by expansion. If it has, we need to restart and write to the new array.
-            if (_newHashTable != hashTableLocal)
+            if (Interlocked.CompareExchange(ref _newHashTable, hashTableLocal, hashTableLocal) != hashTableLocal)
```

Comment:
>Interlocked.CompareExchange [](start = 16, length = 27)

this could be ordinary read/compare.  `TryWriteSentinelToLocation` used full fence to write, so the value you read here will be newer than that write. 
If expansion started before that write, you will see the new table. 
Both writes are full fences, so there is a total order. 

I.E. - it is important that sentinel and new table publishing are full fences, but this read could be a regular read.